File: rs/src/fileobj.rs
//! Buffered writers for [`Fs`] and [`Batch`].
//!
//! [`FsWriter`] accumulates writes and commits on [`close()`](FsWriter::close).
//! [`BatchWriter`] accumulates writes and stages to a [`Batch`] on
//! [`close()`](BatchWriter::close).
//!
//! Both implement [`std::io::Write`] so you can use `write!` / `write_all`.

use std::io;

use crate::batch::Batch;
use crate::error::Result;
use crate::fs::{Fs, WriteOptions};

/// Buffered writer that commits to an [`Fs`] on close.
///
/// Implements [`std::io::Write`] for streaming data. Call
/// [`close()`](FsWriter::close) to flush the buffer and commit.
///
/// # Example
///
/// ```rust,no_run
/// use std::io::Write;
/// use gitstore::{GitStore, Fs};
///
/// let store = GitStore::open("/tmp/repo").unwrap();
/// let fs = store.branches().get("main").unwrap();
/// let mut w = fs.writer("output.bin").unwrap();
/// w.write_all(b"chunk 1").unwrap();
/// w.write_all(b"chunk 2").unwrap();
/// let fs2 = w.close().unwrap();
/// ```
pub struct FsWriter {
    fs: Fs,
    path: String,
    buf: Vec<u8>,
    closed: bool,
}

impl FsWriter {
    pub(crate) fn new(fs: Fs, path: String) -> Self {
        Self {
            fs,
            path,
            buf: Vec::new(),
            closed: false,
        }
    }

    /// Whether this writer has been closed.
    pub fn closed(&self) -> bool {
        self.closed
    }

    /// Flush the buffer, commit, and return the new [`Fs`] snapshot.
    ///
    /// After closing, the writer cannot be used again.
    pub fn close(&mut self) -> Result<Fs> {
        if self.closed {
            return Ok(self.fs.clone());
        }
        let data = std::mem::take(&mut self.buf);
        let new_fs = self.fs.write(&self.path, &data, WriteOptions::default())?;
        self.fs = new_fs.clone();
        self.closed = true;
        Ok(new_fs)
    }
}

impl io::Write for FsWriter {
    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
        if self.closed {
            return Err(io::Error::new(
                io::ErrorKind::Other,
                "I/O operation on closed writer",
            ));
        }
        self.buf.extend_from_slice(buf);
        Ok(buf.len())
    }

    fn flush(&mut self) -> io::Result<()> {
        Ok(())
    }
}

impl Drop for FsWriter {
    fn drop(&mut self) {
        if !self.closed {
            let _ = self.close();
        }
    }
}

/// Buffered writer that stages to a [`Batch`] on close.
///
/// Implements [`std::io::Write`] for streaming data. Call
/// [`close()`](BatchWriter::close) to flush the buffer and stage to the batch.
///
/// # Example
///
/// ```rust,no_run
/// use std::io::Write;
/// use gitstore::{GitStore, Fs};
/// use gitstore::fs::BatchOptions;
///
/// let store = GitStore::open("/tmp/repo").unwrap();
/// let fs = store.branches().get("main").unwrap();
/// let mut batch = fs.batch(Default::default());
/// let mut w = batch.writer("data.bin").unwrap();
/// w.write_all(b"chunk").unwrap();
/// w.close().unwrap();
/// let fs2 = batch.commit().unwrap();
/// ```
pub struct BatchWriter<'a> {
    batch: &'a mut Batch,
    path: String,
    buf: Vec<u8>,
    closed: bool,
}

impl<'a> BatchWriter<'a> {
    pub(crate) fn new(batch: &'a mut Batch, path: String) -> Self {
        Self {
            batch,
            path,
            buf: Vec::new(),
            closed: false,
        }
    }

    /// Whether this writer has been closed.
    pub fn closed(&self) -> bool {
        self.closed
    }

    /// Flush the buffer and stage the write to the batch.
    pub fn close(&mut self) -> Result<()> {
        if self.closed {
            return Ok(());
        }
        let data = std::mem::take(&mut self.buf);
        self.batch.write(&self.path, &data)?;
        self.closed = true;
        Ok(())
    }
}

impl io::Write for BatchWriter<'_> {
    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
        if self.closed {
            return Err(io::Error::new(
                io::ErrorKind::Other,
                "I/O operation on closed writer",
            ));
        }
        self.buf.extend_from_slice(buf);
        Ok(buf.len())
    }

    fn flush(&mut self) -> io::Result<()> {
        Ok(())
    }
}

impl Drop for BatchWriter<'_> {
    fn drop(&mut self) {
        if !self.closed {
            let _ = self.close();
        }
    }
}

File: ts/src/fileobj.ts
/**
 * Buffered writers for FS and Batch.
 *
 * FsWriter accumulates chunks and commits on close().
 * BatchWriter accumulates chunks and stages to a Batch on close().
 */

import type { FS } from './fs.js';
import type { Batch } from './batch.js';

function concatChunks(chunks: Uint8Array[]): Uint8Array {
  if (chunks.length === 0) return new Uint8Array(0);
  if (chunks.length === 1) return chunks[0];
  const total = chunks.reduce((n, c) => n + c.length, 0);
  const out = new Uint8Array(total);
  let offset = 0;
  for (const c of chunks) {
    out.set(c, offset);
    offset += c.length;
  }
  return out;
}

/**
 * Buffered writer that commits to an FS on close.
 *
 * Accepts `Uint8Array` or `string` via `write()`. Strings are UTF-8 encoded.
 *
 * @example
 * ```ts
 * const w = fs.writer('output.bin');
 * w.write(new Uint8Array([1, 2, 3]));
 * await w.close();
 * const newFs = w.fs!;
 * ```
 */
export class FsWriter {
  private _fs: FS;
  private _path: string;
  private _chunks: Uint8Array[] = [];
  private _closed = false;

  /** The resulting FS snapshot after close. Null until `close()` completes. */
  fs: FS | null = null;

  /** @internal */
  constructor(fs: FS, path: string) {
    this._fs = fs;
    this._path = path;
  }

  /** Whether this writer has been closed. */
  get closed(): boolean {
    return this._closed;
  }

  /**
   * Buffer data for writing.
   *
   * @param data - Bytes or string to write. Strings are UTF-8 encoded.
   * @throws {Error} If the writer has been closed.
   */
  write(data: Uint8Array | string): void {
    if (this._closed) throw new Error('I/O operation on closed writer.');
    if (typeof data === 'string') data = new TextEncoder().encode(data);
    this._chunks.push(data);
  }

  /**
   * Flush buffered data and commit.
   *
   * After close, the resulting snapshot is available via `fs`.
   */
  async close(): Promise<void> {
    if (!this._closed) {
      this.fs = await this._fs.write(this._path, concatChunks(this._chunks));
      this._closed = true;
    }
  }
}

/**
 * Buffered writer that stages to a Batch on close.
 *
 * Accepts `Uint8Array` or `string` via `write()`. Strings are UTF-8 encoded.
 *
 * @example
 * ```ts
 * const batch = fs.batch();
 * const w = batch.writer('data.bin');
 * w.write(chunk1);
 * w.write(chunk2);
 * await w.close();
 * const result = await batch.commit();
 * ```
 */
export class BatchWriter {
  private _batch: Batch;
  private _path: string;
  private _chunks: Uint8Array[] = [];
  private _closed = false;

  /** @internal */
  constructor(batch: Batch, path: string) {
    this._batch = batch;
    this._path = path;
  }

  /** Whether this writer has been closed. */
  get closed(): boolean {
    return this._closed;
  }

  /**
   * Buffer data for writing.
   *
   * @param data - Bytes or string to write. Strings are UTF-8 encoded.
   * @throws {Error} If the writer has been closed.
   */
  write(data: Uint8Array | string): void {
    if (this._closed) throw new Error('I/O operation on closed writer.');
    if (typeof data === 'string') data = new TextEncoder().encode(data);
    this._chunks.push(data);
  }

  /**
   * Flush buffered data and stage to the batch.
   */
  async close(): Promise<void> {
    if (!this._closed) {
      await this._batch.write(this._path, concatChunks(this._chunks));
      this._closed = true;
    }
  }
}
